syntax = "proto3";
// -----------------------------------------------------------------------------
// Compilation
// -----------------------------------------------------------------------------

// Install the Google Protobuf protoc compiler:
//   Web: http://code.google.com/p/protobuf/
//   Apt: Search your repository for 'protobuf-compiler'
//
// Compile the schema file to generate source code:
//   C++ : $ protoc --cpp_out=path/to/generated/code  tira_host.proto
//   Java: $ protoc --java_out=path/to/generated/code tira_host.proto
//   Java gRPC: $ protoc --plugin=protoc-gen-grpc-java=build/exe/java_plugin/protoc-gen-grpc-java --grpc-java_out=. --proto_path=. tira_host.proto
//     See https://github.com/grpc/grpc-java/tree/master/compiler
//
// Python:
//   Install gRPC for Python: https://grpc.io/docs/languages/python/quickstart/
//   $ python -m grpc_tools.protoc -I=. --python_out=. --grpc_python_out=. ./tira_host.proto

// -----------------------------------------------------------------------------
// Options
// -----------------------------------------------------------------------------

package tira.generated;

import "google/protobuf/empty.proto";
// import "tira_messages.proto";

// Becomes the package name of the generated Java code
option java_package = "de.webis.tira.client.web.generated";

// option java_multiple_files = true;
option java_outer_classname = "TiraHostMessages";
// option objc_class_prefix = "THM";

// Other options
option optimize_for = SPEED;

// -----------------------------------------------------------------------------
// Definitions
// -----------------------------------------------------------------------------

// The service definition.
service TiraHostService {
  rpc vm_backup (RequestVmCommands) returns (Transaction) {}
  rpc vm_create (RequestVmCreate) returns (Transaction) {}
  rpc vm_delete (RequestVmCommands) returns (Transaction) {}
  rpc vm_info (RequestVmCommands) returns (VmInfo) {}
  rpc vm_list (google.protobuf.Empty) returns (Transaction) {}
  rpc vm_metrics (RequestVmCommands) returns (Transaction) {}
  rpc vm_sandbox (RequestVmCommands) returns (Transaction) {}
  rpc vm_shutdown (RequestVmCommands) returns (Transaction) {}
  rpc vm_snapshot (RequestVmCommands) returns (Transaction) {}
  rpc vm_start (RequestVmCommands) returns (Transaction) {}
  rpc vm_stop (RequestVmCommands) returns (Transaction) {}
  rpc vm_unsandbox (RequestVmCommands) returns (Transaction) {}
  rpc run_execute(RequestRunExecuteEval) returns (Transaction) {}
  rpc run_eval(RequestRunExecuteEval) returns (Transaction) {}
  rpc alive (google.protobuf.Empty) returns (google.protobuf.Empty) {}
}

service TiraApplicationService {
  rpc set_state (SetVmState) returns (Transaction) {}
  rpc complete_transaction (Transaction) returns (Transaction) {}
}

enum Status {
  SUCCESS = 0;
  FAILED = 1;
}

enum VmState {
    UNDEFINED = 0;
    RUNNING = 1;
    POWERED_OFF = 2;
    POWERING_ON = 3;
    POWERING_OFF = 4;
    SANDBOXING = 5;
    UNSANDBOXING = 6;
    EXECUTING = 7;  // may be called sandboxed?
    ARCHIVED = 8;
}

message RequestVmCommands {
  string vmId = 1;
}

message RequestVmCreate {
  string ovaFile = 1;
  string userId = 2;
  string bulkCommandId = 3;
}

message RequestRunExecuteEval {
  string submissionFile = 1;
  string inputDatasetId = 2;
  string inputRunPath = 3;
  string outputDirName = 4;
  string sandboxed = 5;
  string runId = 6;
  string snapshotName = 7;
  string optionalParameters = 8;
}

message Transaction {
  Status status = 1;
  string transactionId = 2;
}

message SetVmState {
  Status status = 1;
  VmState state = 2;
  string vmId = 3;

}

message VmInfo {
  Status status = 1;
  string guestOs = 2;
  string memorySize = 3;
  string numberOfCpus = 4;
  string sshPort = 5;
  string rdpPort = 6;
  string host = 7;
  bool sshPortStatus = 8;
  bool rdpPortStatus = 9;
  VmState state = 10;
}


// TODO: add response message for vm_create command parsing vm access details
//message ResponseVmCreate {
//  string userId = 1;
//  string userPwd = 2;
//}

message CommandState {
  string hostname = 1;

  message Command {
    string id = 1;
    string commandString = 2;
    string startTime = 3;
    string endTime = 4;
    enum Status {
      RUNNING = 0;
      SUCCESS = 1;
      FAILED = 2;
    }
    Status status = 5;
    string logFile = 6;
    int32 returnCode = 7;
    string bulkCommandId = 8;
  }

  repeated Command commands = 2;
}
